sw/device/lib/dif/dif_hmac.c

```diff
@@ -227,48 +200,97 @@ dif_hmac_result_t dif_hmac_irq_force(const dif_hmac_t *hmac,
   return kDifHmacOk;
 }
 
-dif_hmac_result_t dif_hmac_mode_hmac_start(const dif_hmac_t *hmac,
-                                           const uint8_t *key) {
+dif_hmac_result_t dif_hmac_mode_hmac_start(
+    const dif_hmac_t *hmac, const uint8_t *key,
+    const dif_hmac_transaction_config_t config) {
   if (hmac == NULL || key == NULL) {
     return kDifHmacBadArg;
   }
 
-  // Disable SHA256 while configuring and to clear the digest.
-  mmio_region_nonatomic_clear_bit32(hmac->base_addr, HMAC_CFG_REG_OFFSET,
-                                    HMAC_CFG_SHA_EN);
+  // Read current CFG register value.
+  uint32_t device_config =
+      mmio_region_read32(hmac->base_addr, HMAC_CFG_REG_OFFSET);
 
   // Set the HMAC key.
   // TODO Static assert register layout.
   mmio_region_memcpy_to_mmio32(hmac->base_addr, HMAC_KEY_0_REG_OFFSET, key,
                                HMAC_PARAM_NUMWORDS * sizeof(uint32_t));
 
+  // Set the byte-order of the input message.
+  switch (config.message_endianness) {
+    case kDifHmacEndiannessBig:
+    case kDifHmacEndiannessLittle:
+      device_config = bitfield_bit32_write(
+          device_config, HMAC_CFG_ENDIAN_SWAP,
+          config.message_endianness == kDifHmacEndiannessBig);
+      break;
+    default:
+      return kDifHmacError;
+  }
+
+  // Set the byte-order of the digest.
+  switch (config.digest_endianness) {
+    case kDifHmacEndiannessBig:
+    case kDifHmacEndiannessLittle:
+      device_config = bitfield_bit32_write(
+          device_config, HMAC_CFG_DIGEST_SWAP,
+          config.digest_endianness == kDifHmacEndiannessBig);
```

Comment:
Personally I would rather see `device_config` get set in each switch case rather than doing a nested equality check on `config.digest_endianness`.

Reply:
I would say how it's currently written more closely connects the semantics of digest swap, and means you don't 'lose' the `true`/`false` difference between the two sides, but that's also an opinion, and I'm happy to rewrite this.

I might actually pull the `device_config`/`dif_hmac_transaction_config_t` stuff into its own static function so it's not repeated as much?